include/boost/geometry/algorithms/detail/envelope/box.hpp

```diff
@@ -40,36 +42,124 @@ namespace detail { namespace envelope
 {
 
 
-struct envelope_box_on_spheroid
+template
+<
+    std::size_t Index,
+    std::size_t Dimension,
+    std::size_t DimensionCount,
+    typename CS_Tag
+>
+struct envelope_indexed_box
 {
-    template<typename BoxIn, typename BoxOut>
+    template <typename BoxIn, typename BoxOut>
     static inline void apply(BoxIn const& box_in, BoxOut& mbr)
     {
-        BoxIn box_in_normalized = detail::return_normalized<BoxIn>(box_in);
+        detail::indexed_point_view<BoxIn const, Index> box_in_corner(box_in);
+        detail::indexed_point_view<BoxOut, Index> mbr_corner(mbr);
+
+        detail::conversion::point_to_point
+            <
+                detail::indexed_point_view<BoxIn const, Index>,
+                detail::indexed_point_view<BoxOut, Index>,
+                Dimension,
+                DimensionCount
+            >::apply(box_in_corner, mbr_corner);
+    }
+};
 
-        geometry::transform(box_in_normalized, mbr);
+template
+<
+    std::size_t Index,
+    std::size_t DimensionCount
+>
+struct envelope_indexed_box<Index, 0, DimensionCount, spherical_equatorial_tag>
+{
+    template <typename BoxIn, typename BoxOut>
+    static inline void apply(BoxIn const& box_in, BoxOut& mbr)
+    {
+
+       // transform() does not work with boxes of dimension higher
+        // than 2; to account for such boxes we transform the min/max
+        // points of the boxes using the indexed_point_view
+        detail::indexed_point_view<BoxIn const, Index> box_in_corner(box_in);
+        detail::indexed_point_view<BoxOut, Index> mbr_corner(mbr);
+
+        // first transform the units
+        transform_units(box_in_corner, mbr_corner);
```

Comment:
You are right. The default version would not transform units. This is indeed a "problem".

I am still not sure what see as a problem here (besides the special treatment of `Dimension==1`). The idea is the following: given `Dimension` and `DimensionCount`, the code here computes the envelope for all coordinates from `Dimension` (inclusive) to `DimensionCount` (not inclusive). When `Dimension==0` and the coordinate system is the spherical equatorial or geographic one, there exists a specialization. The specialization by itself does not mean that things will be computed many times....